docs/csp.md

```diff
@@ -0,0 +1,118 @@
+# Setting up Content Security Policy with JSS
+
+Content Security Policy (CSP) is a way of whitelisting what resources the browser should allow to load (and reach out to). There are many good resouces on CSP for more information. [This explanation](https://helmetjs.github.io/docs/csp/) is a good place to start.
+
+In the case of JSS we need to set the `style-src` CSP directive. We don't want to just set it to `unsafe-inline` which will allow everything.
+
+The solution is to include a _nonce_:
+
+[MDN/CSP/style-src](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src) notes the following:
+
+> 'nonce-{base64-value}'
+> A whitelist for specific inline scripts using a cryptographic nonce (number used once). **The server must generate a unique nonce value each time it transmits a policy. It is critical to provide an unguessable nonce, as bypassing a resource’s policy is otherwise trivial**. See [unsafe inline](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#Unsafe_inline_script) script for an example.
+
+A conundrum for now is how to generate a unique nonce value for every page visit, as webpack must know the nonce at bundle time, while the server must know the nonce value at runtime for including it in the CSP rules. For this tutorial, we'll use a hard-coded Base64-encoded GUID value.
+
+1. Generate a Base64 guid:
+
+   * Get a guid: https://www.guidgenerator.com/online-guid-generator.aspx
+   * Encode the guid: https://www.base64encode.org/
+
+1. In server startup:
+
+```js
+// server.js
+import helmet from "helmet";
+
+const WEBPACK_NONCE = "N2M0MDhkN2EtMmRkYi00MTExLWFhM2YtNDhkNTc4NGJhMjA3";
+
+app.use(
+  helmet.contentSecurityPolicy({
+    directives: {
+      /* ... */
+      styleSrc: ["'self'", `'nonce-${WEBPACK_NONCE}'`]
+    }
+  })
+);
+```
+
+(The above example uses [Helmet](https://helmetjs.github.io/) to set CSP directives).
+
+When loading the initial index page, the `Content-Security-Policy` header should now contain the nonce:
+
+```
+style-src 'self' 'nonce-N2M0MDhkN2EtMmRkYi00MTExLWFhM2YtNDhkNTc4NGJhMjA3'
+```
+
+2. In webpack config:
+
+Set the `global.__webpack_nonce__` variable to the same nonce value:
+
+```js
+// webpack.config.js
+const WEBPACK_NONCE = 'N2M0MDhkN2EtMmRkYi00MTExLWFhM2YtNDhkNTc4NGJhMjA3'
+module.exports = {
+  /* ... */
+  plugins: {
+    new webpack.DefinePlugin({
+      'global.__webpack_nonce__': JSON.stringify(WEBPACK_NONCE),
```

Comment:
oh right, it adds additional quotes

Reply:
Yeah it's ugly, but the DefinePlugin is a weird beast. From their Usage section:

> - If the value is a string it will be used as a code fragment.
> - If the value isn't a string, it will be stringified (including functions).

They also use `JSON.stringify` in their examples: https://webpack.js.org/plugins/define-plugin/